Feature/fixes

tgouala opened the pull request:
2 fixes.
Still need one regression test I will write later.

Would love to have these in develop and then make a new release of Guacamole to test it in .Peek.

Cheers,
Thibault.

WellFiredDevelopment commented:
Good catch with 11c56fc6a782f8b6d81fbaeebe9a1a05f3b97783, could you add the regression test from your todo?

Also, I'm not 100% sure from your description what the bug is with 11c56fc6a782f8b6d81fbaeebe9a1a05f3b97783. You gave an example : For Example, TextEntry.TextProperty, but it doesn't really help me understand the problem.

Also, reset your commit to before the merge and do a rebase, you can use your desktop client if you can't on the github website. Alternatively, push a commit with the fix and i'll sort the rebase step later. Lets try to keep the timeline linear.

tgouala commented:
> reset your commit to before the merge and do a rebase

So about the merge, it was quick with Github and since we will delete the branch later, all these merges will disappear (since we merge in the feature branch, not in develop). But I don't mind doing it from client app as long as github detect it.

Note that when you finish the review, you can on github squash commit on top of develop, and in one click delete feature branch, so at the end, everything is linear.

>Also, I'm not 100% sure from your description what the bug is with 11c56fc

In BindableObject, when you set the BindingContext, I am enumerating the different BindableContexts from the dictionary `Dictionary<BindableProperty, BindableContext> _contexts`.

The problem is that new bindable contexts are added in a lazy manner. They are added only if they don't exist when we need it (`GetOrCreateBindableContext(bindableProperty)`).

So with TextEntry, here what was happening :
1. We set the BindingContext of TextEntry, 
2. We go through all the BindableContext (`Dictionary<BindableProperty, BindableContext> _contexts`) and update the value from source. This is a mistake I introduced previously and I am fixing in the current commit [11c56fc](https://github.com/WellFiredDevelopment/wellfired.guacamole/commit/11c56fc6a782f8b6d81fbaeebe9a1a05f3b97783)
3. Text property was already accessed so the BindableContext for TextProperty exists. But when we update the value of Text property, then HorizontalTextAlign and VerticalTextAlign properties getter are accessed (see TextEntryViewRenderer.cs)
4. Since it is the first time these getter are accessed, a BindableContext is created for each of them

===> The consequence is `Dictionary<BindableProperty, BindableContext> _contexts` get updated while we are enumerating it.

So I just introduced a list of all the bound properties and enumerate them instead of enumerating all the Bindable Contexts.

Note that before I simplify you were enumerating this dictionary : `_bindings : <string propertyName, BindableProperty>`  which is equivalent (but I don't think we need a dictionary for that for the moment, so simply use a list).
